File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps, visited, m, n):
    q = deque([(0, 0, 0)]) # (x, y, level)
    visited[0][0] = True
    
    dy = [-1, 1, 0, 0]
    dx = [0, 0, -1, 1]
    
    min_level = 1e9
    is_arrived = False
    
    while (q):
        # print(f"queue = {q}")
        vx, vy, vl = q.popleft()
        
        for i in range(4):
            y = vy + dy[i]
            x = vx + dx[i]
            
            # print(f"x = {x}, y = {y}")
            
            if (x < 0 or x >= m or y < 0 or y >= n):
                continue
            
            if (maps[y][x] == 0):
                continue
                
            if not visited[y][x]:
                q.append((x, y, vl + 1))
                visited[y][x] = True
                
            if x == m-1 and y == n-1:
                min_level = min(min_level, vl + 1)
                is_arrived = True
                
    if is_arrived:
        return min_level + 1 # node 수 = level + 1
    
    return -1

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    visited = [[False for _ in range(m)] for _ in range(n)]
    return bfs(maps, visited, m, n)

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    class Node {
        int r, c;
        int level;
        Node(int r, int c, int l) {
            this.r = r;
            this.c = c;
            this.level = l;
        }
    }
    
    int[][] map;
    int n, m;
    int[] dr = {-1, 1, 0, 0};
    int[] dc = {0, 0, -1, 1};
    
    public int solution(int[][] maps) {
        map = maps;
        n = map.length;
        m = map[0].length;
        
        return bfs();
    }
    
    public int bfs() {
        Deque<Node> q = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][m];
        
        q.add(new Node(0, 0, 0));
        visited[0][0] = true;
        
        while (!q.isEmpty()) {
            Node node = q.poll();
            if (node.r == n-1 && node.c == m-1) {
                return node.level + 1;
            }
            
            for (int i = 0; i < 4; i++) {
                int nr = node.r + dr[i];
                int nc = node.c + dc[i];
                
                if (!isValid(nr, nc)) continue;
                if (visited[nr][nc]) continue;
                if (map[nr][nc] == 0) continue;
                
                visited[nr][nc] = true;
                q.add(new Node(nr, nc, node.level + 1));
            }
        }
        
        return -1;
    }
    
    public boolean isValid(int r, int c) {
        return (r >= 0 && r < n && c >= 0 && c < m);
    }
}